_posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md

```diff
@@ -0,0 +1,196 @@
+
+We decided to check how that really impacts the write queries to the database so we made a little experiment. We ran the insert of 500 entries with ORM, without ORM and using UNWIND:
+
+{% highlight console %}
+Started writing 500 entries with ORM.
+ 500/500 [============================] 100%
+Started in 28 seconds.
+Starting writing 500 entries with connection.
+ 500/500 [============================] 100%
+Finished in 5 seconds.
+Started writing 500 entries with connection and UNWIND.
+Finished in 0 seconds.
+{% endhighlight %}
+
+The write with ORM took 28, using CYPHER 5, using UNWIND took below 1s!
+
+##The solution
+
+When checking the messages stuck on the queue we noticed one message (`AdvertServedToUser`) was 80% of the total messages. So we decided to move it out of ORM and use UNWIND. 
+We created another queue for our batch messages, and when consuming them from the original queue – we just re-queued them to another. We were using [RabbitMqBundle](https://github.com/php-amqplib/RabbitMqBundle) for consuming messages from the queue and Symfony Messenger inside application.
+
+We had to change routing for the specific message in order to put it on the different queue:
+
+{% highlight yaml %}
+#config/packages/messenger.yaml
+framework:
+    messenger:
+        transports:
+            batch_messages_amqp: '%env(BATCH_MESSAGES_TRANSPORT_DSN)%/advert-served-to-user
+        ...
+        routing:
+            App\Message\Command\AdvertServedToUser: batch_messages_amqp
+
+{% endhighlight %}
+
+The next step was to get messages from the new queue in a batch of 50 or more (we ended up getting 5000 of them in one batch). In order to do that we had to create consumer class:
+
+{% highlight php %}
+<?php
+
+declare(strict_types=1);
+
+namespace App\Consumer;
+
+use App\Message\Command\Advert\AdvertServedToUser;
+use App\Message\Command\Advert\RunAdvertServedToUserBatch;
+use OldSound\RabbitMqBundle\RabbitMq\BatchConsumerInterface;
+use PhpAmqpLib\Message\AMQPMessage;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Messenger\MessageBusInterface;
+
+class AdvertServedToUserToUserBatchConsumer implements BatchConsumerInterface
+{
+    private $commandBus;
+    private $logger;
+
+    public function __construct(MessageBusInterface $commandBus, LoggerInterface $logger)
+    {
+        $this->commandBus = $commandBus;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @param AMQPMessage[] $messages
+     *
+     * @return array|bool
+     */
+    public function batchExecute(array $messages)
+    {
+        try {
+            $command = new RunAdvertServedToUserBatch($messages);
+            $this->commandBus->dispatch($command);
+            return true;
+        } catch (\Throwable $e) {
+            $this->logger->log('error', $e->getMessage(), ['exception' => $e]);
+            return false;
+        } 
+    }
+}
+{% endhighlight %}
+
+registered class as a service:
```

Comment:
Im not sure this code example is needed.

Reply:
we are using service id in  `old_sound_rabbit_mq` `consumer` as a `callback` parameter